File: recoverTree.js





function Solution(){


let first =  null;
let second =  null;
let prev =  {val:-Infinity}
return {recover}

function recover(root){
    traverse(root);
    first.val=[second.val, second.val = first.val][0]; 
  
    return root;
}

function traverse(node){
if(node){
    traverse(node.left);
    if(!first && prev.val >= node.val)first = prev;
    if(first && prev.val >= node.val)second = node;
    prev = node;
    traverse(node.right);
        }
    }
}


const obj = new Solution ()
const a= obj.recover({"val":1,"right":null,"left":{"val":3,"right":{"val":2,"right":null,"left":null},"left":null}})
const arr = [];
let i = 3;
while (i--) arr.push(3 - i);
arr
function toBinary(arr) {
    const serial = [];
    return {
        bind,
        seralize
    };

    function bind(lo = 0, hi = arr.length - 1, node = {}) {
        if (lo > hi) return null;
        const mid = lo + ((hi - lo) >> 1);
        node.val = arr[mid]
        node.left = bind(lo, mid - 1, {})
        
        node.right = bind(mid + 1, hi, {})
        return node;
    }

    function seralize(node) {
        if(!node){ 
            serial.push(null)
            return;
        } else {
          serial.push(node.val);
            seralize(node.left);
              
            seralize(node.right);
        }
        return serial;
    }



}



let tree = new toBinary(arr);

let root = tree.bind()




let node = tree.seralize(root)

node

File: recoverTree.cpp
#include<stack>
#include <iostream>
#include<bits/stdc++.h> 
using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(NULL), right(NULL){}
};

class Solution{
 
    TreeNode* first = NULL;
    TreeNode* second = NULL;
    TreeNode* prev = new TreeNode(INT_MIN);
public:

    void recoverTree(TreeNode* root) {
        help(root);
        swap(first->val, second->val);
    }
    
    void help(TreeNode* node){
        if(!node)  return;
        help(node->left);
        if(!first && prev->val >= node -> val)   first = prev;
        if(first && prev->val >= node -> val)   second = node;
        prev = node;
        help(node -> right);
    }


};
class SolutionTwo {
private:
    TreeNode* pre=NULL;
    TreeNode* first=NULL;
    TreeNode* second=NULL;
    
    void  inOrder(TreeNode * root){
        if(!root) return;
        inOrder(root->left);
        if(pre!=NULL && first == NULL && pre-> val >= root-> val){
            first = pre;
        }
        
        if(first != NULL && pre-> val >= root -> val){
            second = root;
            if(pre != first){
                return;
            }
        }
        pre=root;
        inOrder(root->right);
    }
    
public:
    void recoverTree1(TreeNode* root) {
        stack<TreeNode*> helper;
        TreeNode* pre=NULL;
        bool first=false;
        TreeNode *second;
        int tag=INT_MAX;
        while(root!=NULL || !helper.empty()){
            while(root != NULL){
                helper.push(root);
                root = root->left;
            }
            root=helper.top();
            helper.pop();
            if(pre != NULL && !first && pre->val >= root -> val){
                first = true;
                second = root;
                tag=root -> val;
            }
            if(!first){
                pre=root;
            }else if(root -> val < tag){
                tag=root -> val;
                second = root;
            }
            
            root= root -> right;
        }
        int tmp= pre -> val;
        pre -> val = second -> val;
        second -> val = tmp;
    }
    
    void recoverTree(TreeNode* root) {
        inOrder(root);
        int tmp= first-> val;
        first -> val= second-> val;
        second -> val = tmp;
    }




};


